test/e2e/apimachinery/validatingadmissionpolicy.go

```diff
@@ -354,6 +354,33 @@ var _ = SIGDescribe("ValidatingAdmissionPolicy [Privileged:ClusterAdmin]", frame
 			gomega.Expect(warning.FieldRef).To(gomega.Equal("spec.validations[1].expression"))
 			gomega.Expect(warning.Warning).To(gomega.ContainSubstring("undefined field 'maxRetries'"))
 		})
+		ginkgo.By("adding maxRetries as an integer field to the CRD.", func() {
```

Comment:
I feel like we can get more extensive tests than this. This only demonstrates a single update. I'd have preferred to see a bit more of an effort to trip up the controller.

Also curious to know if integration test is possible if we run the controller in the test directly against the testserver?

Reply:
It would be great if we can defer it to the test freeze. And yes we should prefer lower-level tests. The E2E test may wait for up to 6 seconds (the pull interval), if the mutation happens right after a pull. It will quickly add up, so I believe it is better to leave the e2e test brief here.